src/btree/bt_read.c

```diff
@@ -118,24 +117,65 @@ __page_read(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags)
+     * Don't do this for deleted pages, as the reconciliation needs to examine the page delete
+     * information. That requires locking the ref, which requires waiting for the read to finish.
+     * (It is possible that always writing out a reference to the on-disk version of the page is
+     * sufficient in this case, but it's not entirely clear; we expect reads of deleted pages to be
+     * rare, so it's better to do the safe thing.)
      */
     if (previous_state == WT_REF_DISK)
         F_SET(ref, WT_REF_FLAG_READING);
 
     /*
      * Get the address: if there is no address, the page was deleted and a subsequent search or
-     * insert is forcing re-creation of the name space.
+     * insert is forcing re-creation of the name space. There can't be page delete information,
+     * because that information is an amendment to an on-disk page; when a page is deleted any page
+     * delete information should expire and be removed before the original on-disk page is actually
+     * discarded.
      */
     if (!__wt_ref_addr_copy(session, ref, &addr)) {
         WT_ASSERT(session, previous_state == WT_REF_DELETED);
-
+        WT_ASSERT(session, ref->page_del == NULL);
         WT_ERR(__wt_btree_new_leaf_page(session, ref));
         goto skip_read;
     }
 
+    /*
+     * If the page is deleted and the deletion is globally visible, don't bother reading and
+     * explicitly instantiating the existing page. Get a fresh page and pretend we got it by reading
+     * the on-disk page. Note that it's important to set the instantiated flag on the page so that
+     * reconciling the parent internal page knows it was previously deleted. Otherwise it's possible
+     * to write out a reference to the original page without the deletion, which will cause it to
+     * come back to life unexpectedly.
+     *
+     * Setting the instantiated flag requires a modify structure. We don't need to mark it dirty; if
+     * it gets discarded before something else modifies it, eviction will see the instantiated flag
+     * and set the ref state back to WT_REF_DELETED.
+     *
+     * Skip this optimization in cases that need the obsolete values. To minimize the number of
+     * special cases, use the same test as for skipping instantiation below.
+     */
+    if (previous_state == WT_REF_DELETED &&
+      !F_ISSET(S2BT(session), WT_BTREE_SALVAGE | WT_BTREE_UPGRADE | WT_BTREE_VERIFY)) {
+        /*
+         * If the deletion has not yet been found to be globally visible (page_del isn't NULL),
+         * check if it is now, in case we can in fact avoid reading the page. Hide prepared deletes
+         * from this check; if the deletion is prepared we still need to load the page, because the
+         * reader might be reading at a timestamp early enough to not conflict with the prepare.
+         */
+        if (ref->page_del != NULL && __wt_page_del_visible_all(session, ref->page_del, true))
```

Comment:
Should we update the oldest id to make it more likely that the page_del is globally visible?

Reply:
Good idea. Hadn't thought of that. Added in [13f7c10](https://github.com/wiredtiger/wiredtiger/pull/8188/commits/13f7c10a82cf5980908c3540c1766f4108cc81c1)